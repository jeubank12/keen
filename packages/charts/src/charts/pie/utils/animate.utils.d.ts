import { MutableRefObject } from 'react';
import { Arc, DefaultArcObject } from 'd3-shape';
import 'd3-transition';
import { ValueFn } from 'd3-selection';
export declare type ArcProperties = {
  startAngle: number;
  endAngle: number;
};
export declare const createArcTween: (
  currentArc: ArcProperties,
  updatedArc: ArcProperties,
  arc: Arc<any, Partial<DefaultArcObject>>
) => () => (t: number) => string;
export declare const animateArcPath: (
  element: MutableRefObject<SVGPathElement>,
  tween: ValueFn<any, any, (t: number) => string>,
  onMotionEnd: () => void,
  duration?: number
) => void;
