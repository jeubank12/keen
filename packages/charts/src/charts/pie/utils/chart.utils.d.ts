import { Dimension, Margins, DataSelector } from '../../../types';
export declare type LabelsPosition = 'inside' | 'outside';
export declare type Options = {
  data: Record<string, any>[];
  labelSelector: string;
  keys: string[];
  disabledLabels: string[];
  dimension: Dimension;
  margins: Margins;
  colors: string[];
  padAngle: number;
  padRadius: number;
  cornerRadius: number;
  innerRadius: number;
  labelsRadius: number;
  labelsPosition: LabelsPosition;
};
declare type PieSlice = {
  index: string;
  label: string;
  activePosition: [number, number];
  labelPosition: [number, number];
  color: string;
  selector: DataSelector;
  startAngle: number;
  endAngle: number;
};
export declare const HOVER_RADIUS = 5;
export declare const generatePieChart: ({
  data,
  colors,
  dimension,
  padAngle,
  padRadius,
  cornerRadius,
  innerRadius,
  labelSelector,
  keys,
  disabledLabels,
  labelsPosition,
  labelsRadius,
  margins,
}: Options) => {
  arcs: PieSlice[];
  drawActiveArc: import('d3-shape').Arc<
    any,
    import('d3-shape').DefaultArcObject
  >;
  drawArc: import('d3-shape').Arc<any, import('d3-shape').DefaultArcObject>;
};
export {};
