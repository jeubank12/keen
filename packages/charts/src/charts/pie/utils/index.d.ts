import { createArcTween, animateArcPath, ArcProperties } from './animate.utils';
import { getTooltipContent } from './tooltip.utils';
import {
  generatePieChart,
  LabelsPosition,
  Options as PieChartOptions,
} from './chart.utils';
export {
  generatePieChart,
  getTooltipContent,
  createArcTween,
  animateArcPath,
  ArcProperties,
  LabelsPosition,
  PieChartOptions,
};
