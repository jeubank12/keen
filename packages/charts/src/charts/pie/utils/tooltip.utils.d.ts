import { DataSelector } from '../../../types';
declare type Options = {
  data: Record<string, any>[];
  keys: string[];
  selectors: {
    selector: DataSelector;
    color: string;
  }[];
};
export declare const getTooltipContent: ({
  data,
  keys,
  selectors,
}: Options) => {
  name: string;
  value: number;
}[];
export {};
