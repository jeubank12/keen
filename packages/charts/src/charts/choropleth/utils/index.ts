import { fetchMapTopology, MapType } from './map.utils';
import { generateChoropleth } from './chart.utils';

export { fetchMapTopology, generateChoropleth, MapType };
