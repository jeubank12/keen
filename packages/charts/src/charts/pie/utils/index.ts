import {
  generatePieChart,
  LabelsPosition,
  Options as PieChartOptions,
} from './chart.utils';

export { generatePieChart, LabelsPosition, PieChartOptions };
