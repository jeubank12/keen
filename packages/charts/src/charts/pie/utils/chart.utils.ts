import { sum } from 'd3-array';
import { arc, pie } from 'd3-shape';

import { calculateHypotenuse } from '../../../utils/math.utils';

import { Dimension, Margins, DataSelector } from '../../../types';

export type LabelsPosition = 'inside' | 'outside';

export type Options = {
  data: Record<string, any>[];
  labelSelector: string;
  keys: string[];
  disabledLabels: string[];
  dimension: Dimension;
  margins: Margins;
  colors: string[];
  padAngle: number;
  innerRadius: number;
  labelsRadius: number;
  labelsPosition: LabelsPosition;
};

type PieSlice = {
  index: string;
  label: string;
  labelPosition: [number, number];
  color: string;
  selector: DataSelector;
  startAngle: number;
  endAngle: number;
};

type PieValue = { color: string; value: number; selector: DataSelector };

export const HOVER_RADIUS = 10;

export const generatePieChart = ({
  data,
  colors,
  dimension,
  padAngle,
  innerRadius,
  labelSelector,
  keys,
  disabledLabels,
  labelsPosition,
  labelsRadius,
  margins,
}: Options) => {
  const values: PieValue[] = [];

  const { width, height } = dimension;
  const radius =
    Math.min(
      width - margins.left - margins.right,
      height - margins.top - margins.bottom
    ) / 2;

  data.forEach((item, idx) => {
    const label = item[labelSelector];
    if (!disabledLabels.includes(label)) {
      const result = keys.reduce((acc, currentKey) => {
        if (currentKey !== label) return acc + item[currentKey];
        return acc;
      }, 0) as number;

      values.push({
        value: result,
        selector: [idx],
        color: colors[idx],
      });
    }
  });

  const total = sum(values, d => d.value);
  const createPie = pie().value((d: any) => d.value);

  const createArc = arc().padAngle(padAngle);

  const calculateLabelPosition = (
    startAngle: number,
    endAngle: number
  ): [number, number] => {
    const [x, y] = createArc.centroid({
      innerRadius: innerRadius,
      outerRadius: radius + labelsRadius,
      startAngle,
      endAngle,
    });

    if (labelsPosition === 'outside') {
      return calculateHypotenuse(x, y, radius + labelsRadius);
    }

    return [x, y];
  };

  const arcs: PieSlice[] = createPie(values as any).map(
    ({ startAngle, endAngle, value, index, data }) => {
      const { color, selector } = data as any;

      return {
        label: String(`${(Math.round(value * 100) / total).toFixed(1)}%`),
        labelPosition: calculateLabelPosition(startAngle, endAngle),
        index: String(index),
        selector,
        color,
        startAngle,
        endAngle,
      };
    }
  );

  return {
    arcs,
    drawActiveArc: arc()
      .padAngle(padAngle)
      .innerRadius(innerRadius)
      .outerRadius(radius + HOVER_RADIUS),
    drawArc: arc()
      .padAngle(padAngle)
      .innerRadius(innerRadius)
      .outerRadius(radius),
  };
};
