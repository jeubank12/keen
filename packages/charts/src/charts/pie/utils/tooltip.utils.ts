import { getFromPath } from '../../../utils';

import { DataSelector } from '../../../types';

type Options = {
  data: Record<string, any>[];
  keys: string[];
  selectors: { selector: DataSelector; color: string }[];
};

export const getTooltipContent = ({ data, keys, selectors }: Options) => {
  const content: { name: string; value: number }[] = [];

  selectors.forEach(({ selector }) => {
    const item = getFromPath(data, selector);
    keys.forEach((keyName: string) => {
      const value = item[keyName];
      content.push({ name: keyName, value });
    });
  });

  return content;
};
