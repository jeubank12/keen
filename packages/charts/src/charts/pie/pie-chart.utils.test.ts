import { generatePieChart } from './pie-chart.utils';

import { pieChart } from './pie-chart.fixtures';

describe('@keen/charts', () => {
  describe('<PieChart /> - utils', () => {
    it('should create arcs based on provided data', () => {
      const { arcs } = generatePieChart(pieChart);
      const { data } = pieChart;

      expect(arcs.length).toEqual(data.length);
    });

    it('should calculate properties for each arc', () => {
      const { arcs } = generatePieChart(pieChart);

      expect(arcs).toMatchSnapshot();
    });

    it('should calculate position for labels placed "outer" of slice', () => {
      const { arcs } = generatePieChart({
        ...pieChart,
        labelsPosition: 'outside',
      });

      expect(arcs).toMatchSnapshot();
    });
  });
});
