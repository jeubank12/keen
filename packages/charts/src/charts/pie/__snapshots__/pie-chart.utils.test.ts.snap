// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@keen/charts <PieChart /> - utils should calculate position for labels placed "outer" of slice 1`] = `
Array [
  Object {
    "index": "4",
    "label": "14.2%",
    "labelPosition": Array [
      -34.501855139563546,
      -72.1777111851614,
    ],
    "path": "M-30.987399294969656,-25.29389422240304A40,40,0,0,1,-0.2236058660561739,-39.99937500032551L-0.22360586605615085,-19.99874997135234A20,20,0,0,0,-15.422757044954187,-12.733364250358816Z",
    "pathActive": "M-46.493226654345136,-37.92597889663089A60,60,0,0,1,-0.31622644840281067,-59.99916666782407L-0.3162264484027946,-19.997499864566336A20,20,0,0,0,-15.36361783986544,-12.80465723362279Z",
  },
  Object {
    "index": "2",
    "label": "20.6%",
    "labelPosition": Array [
      -44.36132046019484,
      66.57381802952193,
    ],
    "path": "M2.2074716889161894,39.93904191067474A40,40,0,0,1,-37.70994942812798,13.340154201808556L-18.817392212389276,6.775378242217287A20,20,0,0,0,0.99208255813902,19.975379150289896Z",
    "pathActive": "M3.330360763998235,59.90750117624354A60,60,0,0,1,-56.5713187334494,19.992145876780633L-18.785810645042357,6.862457169892807A20,20,0,0,0,0.8995571778492524,19.97975968033099Z",
  },
  Object {
    "index": "0",
    "label": "25.2%",
    "labelPosition": Array [
      56.85445267025227,
      -56.28117990559584,
    ],
    "path": "M0.22360586605615126,-39.99937500032551A40,40,0,0,1,39.999587043065965,0.18175913784315811L19.999989051384123,-0.020927123910416574A20,20,0,0,0,0.22360586605615285,-19.99874997135234Z",
    "pathActive": "M0.31622644840280334,-59.99916666782407A60,60,0,0,1,59.99929033271326,0.2918211280588869L19.99967762543816,-0.11355561874374182A20,20,0,0,0,0.3162264484027922,-19.997499864566336Z",
  },
  Object {
    "index": "3",
    "label": "16.1%",
    "labelPosition": Array [
      -78.81570540969109,
      -13.714393197396307,
    ],
    "path": "M-37.85673708219204,12.91771874170459A40,40,0,0,1,-31.26825133618883,-24.945870567629534L-15.703609086173367,-12.385340595585301A20,20,0,0,0,-18.964179866453346,6.352942782113306Z",
    "pathActive": "M-56.77890782461574,19.394731919874257A60,60,0,0,1,-46.89041142005747,-37.43379912402353L-15.760802605577762,-12.312477461015444A20,20,0,0,0,-18.99339973620869,6.265043212986452Z",
  },
  Object {
    "index": "1",
    "label": "23.9%",
    "labelPosition": Array [
      57.98342297832959,
      55.11735352605493,
    ],
    "path": "M39.99505500098913,0.6289479055168361A40,40,0,0,1,2.6538569488490307,39.91186594604474L1.4384678180718669,19.948203185659892A20,20,0,0,0,19.995457009307295,0.4262616437632615Z",
    "pathActive": "M59.992881057342956,0.9242415482439135A60,60,0,0,1,3.961644852638723,59.86906855849322L1.530841266489723,19.94132706258067A20,20,0,0,0,19.993268350067865,0.5188648014412846Z",
  },
]
`;

exports[`@keen/charts <PieChart /> - utils should calculate properties for each arc 1`] = `
Array [
  Object {
    "index": "4",
    "label": "14.2%",
    "labelPosition": Array [
      -14.375772974818144,
      -30.074046327150587,
    ],
    "path": "M-30.987399294969656,-25.29389422240304A40,40,0,0,1,-0.2236058660561739,-39.99937500032551L-0.22360586605615085,-19.99874997135234A20,20,0,0,0,-15.422757044954187,-12.733364250358816Z",
    "pathActive": "M-46.493226654345136,-37.92597889663089A60,60,0,0,1,-0.31622644840281067,-59.99916666782407L-0.3162264484027946,-19.997499864566336A20,20,0,0,0,-15.36361783986544,-12.80465723362279Z",
  },
  Object {
    "index": "2",
    "label": "20.6%",
    "labelPosition": Array [
      -18.483883525081186,
      27.739090845634138,
    ],
    "path": "M2.2074716889161894,39.93904191067474A40,40,0,0,1,-37.70994942812798,13.340154201808556L-18.817392212389276,6.775378242217287A20,20,0,0,0,0.99208255813902,19.975379150289896Z",
    "pathActive": "M3.330360763998235,59.90750117624354A60,60,0,0,1,-56.5713187334494,19.992145876780633L-18.785810645042357,6.862457169892807A20,20,0,0,0,0.8995571778492524,19.97975968033099Z",
  },
  Object {
    "index": "0",
    "label": "25.2%",
    "labelPosition": Array [
      23.689355279271783,
      -23.4504916273316,
    ],
    "path": "M0.22360586605615126,-39.99937500032551A40,40,0,0,1,39.999587043065965,0.18175913784315811L19.999989051384123,-0.020927123910416574A20,20,0,0,0,0.22360586605615285,-19.99874997135234Z",
    "pathActive": "M0.31622644840280334,-59.99916666782407A60,60,0,0,1,59.99929033271326,0.2918211280588869L19.99967762543816,-0.11355561874374182A20,20,0,0,0,0.3162264484027922,-19.997499864566336Z",
  },
  Object {
    "index": "3",
    "label": "16.1%",
    "labelPosition": Array [
      -32.83987725403796,
      -5.714330498915128,
    ],
    "path": "M-37.85673708219204,12.91771874170459A40,40,0,0,1,-31.26825133618883,-24.945870567629534L-15.703609086173367,-12.385340595585301A20,20,0,0,0,-18.964179866453346,6.352942782113306Z",
    "pathActive": "M-56.77890782461574,19.394731919874257A60,60,0,0,1,-46.89041142005747,-37.43379912402353L-15.760802605577762,-12.312477461015444A20,20,0,0,0,-18.99339973620869,6.265043212986452Z",
  },
  Object {
    "index": "1",
    "label": "23.9%",
    "labelPosition": Array [
      24.159759574303997,
      22.965563969189557,
    ],
    "path": "M39.99505500098913,0.6289479055168361A40,40,0,0,1,2.6538569488490307,39.91186594604474L1.4384678180718669,19.948203185659892A20,20,0,0,0,19.995457009307295,0.4262616437632615Z",
    "pathActive": "M59.992881057342956,0.9242415482439135A60,60,0,0,1,3.961644852638723,59.86906855849322L1.530841266489723,19.94132706258067A20,20,0,0,0,19.993268350067865,0.5188648014412846Z",
  },
]
`;
